### DISCUSSION TITLE
field.visible = false 还保留了值

### DISCUSSION THREAD

**Discussion by rxlf** [alibaba/formily#2580] [Q&A 2.x] [answered]
根据显隐规则：

> 字段的显示隐藏，主要用 display 属性来表达：
> display 为 none 代表字段 UI 隐藏，同时不保留字段数据
> display 为 hidden 代表字段 UI 隐藏，保留字段数据
> display 为 visible 代表字段 UI 显示，同时恢复字段数据
> 在 display 属性之上，我们还提供了两个便捷属性
> 
> visible，如果为 true 代表 display 等于 visible，如果为 false 代表 display 等于 none
> hidden，如果为 true 代表 display 等于 hidden，如果为 false 代表 display 等于 visible

设置 visible 为 false 应该清空值，但是目前没有清空，demo 链接： https://codesandbox.io/s/brave-bassi-iw2fe?file=/App.tsx

**Comment by janryWang** [comment, marked as answer]
你设置的是hidden

**Reply by rxlf** [reply]
const form = createForm({
  effects: () => {
    onFieldReact("radio", (field, form) => {
      const input = form.query("input").take();
      input.visible = field?.value === 1;
    });
  }
});

可能是刚刚没有保存，visible 也是一样的

**Reply by janryWang** [reply]
应该用NumberPicker，Input拿到的值是字符串，这是基础知识

**Reply by rxlf** [reply]
没理解和 NumberPicker 有什么关系
在我的 demo 里，当在 input 里输入任意字符串后，通过 visible 隐藏 input，按照定义应该清空 input 的 value 啊，现在的问题是 Input 的 value 没有清空。。这似乎和定义不太相符

**Reply by janryWang** [reply]
你用的===，是没办法判断等于数字1的，这种基础知识都没理解？

**Reply by rxlf** [reply]
我是判断的 radio 的值啊。。。我 radio 值本来就是 number 啊

场景是这样的，我在 input 里输入了字符串(如输入了 `'aaa'`)，然后改变 radio 的值为 2， 这时候 `Input.visible = false`，input 被隐藏，但是当我设置 radio 值为 1 时，`Input.visible = true`, 同时 Input 的值被还原成了 `'aaa'`，按照定义不应该还原啊

**Reply by janryWang** [reply]
好吧，所以你的问题是隐藏到显示值还原的问题？这个是内置逻辑，就是要恢复现场的，如果你要清空缓存，你可以单独调用field.reset

**Reply by rxlf** [reply]
是的，之前的定义是` visible = false` 的时候会默认清空啊，是改了？

**Reply by janryWang** [reply]
隐藏的时候会删值，重新显示的时候会恢复，逻辑一直都是这样的

**Reply by rxlf** [reply]
那就和 https://core.formilyjs.org/zh-CN/guide/field#%E6%98%BE%E9%9A%90%E8%A7%84%E5%88%99 描述不一样了
visible = false  ， display 就应该为 none ，描述是不保留数据字段，现在是保留了

**Reply by janryWang** [reply]
现在visible为false就是display为none，这个时候提交的时候是不会带值的

**Reply by rxlf** [reply]
👌
